src/ai/backend/client/func/vfolder.py

```diff
@@ -374,15 +371,57 @@ async def upload(
             )
             await uploader.upload()
             input_file.close()
-        return None
+
+    async def _upload_recursively(
+        self,
+        source: Sequence[Path],
+        basedir: Union[str, Path] = None,
+        dst_dir: Union[str, Path] = None,
+        chunk_size: int = DEFAULT_CHUNK_SIZE,
+        address_map: Optional[Mapping[str, str]] = None,
+    ) -> None:
+        dir_list: list[Path] = []
+        file_list: list[Path] = []
+        base_path = Path.cwd() if basedir is None else Path(basedir).resolve()
+        for path in source:
+            if path.is_file():
+                file_list.append(path)
+            else:
+                await self._mkdir(path.relative_to(base_path))
+                dir_list.append(path)
+        await self._upload_files(file_list, basedir, dst_dir, chunk_size, address_map)
+        for dir in dir_list:
+            await self._upload_recursively(
+                list(dir.glob("*")), basedir, dst_dir, chunk_size, address_map
+            )
 
     @api_function
-    async def mkdir(
+    async def upload(
+        self,
+        sources: Sequence[Union[str, Path]],
+        *,
+        basedir: Union[str, Path] = None,
+        recursive: bool = False,
+        dst_dir: Union[str, Path] = None,
+        chunk_size: int = DEFAULT_CHUNK_SIZE,
+        address_map: Optional[Mapping[str, str]] = None,
+        show_progress: bool = False,
+    ) -> None:
+        if basedir:
+            src_paths = [basedir / Path(src) for src in sources]
+        else:
+            src_paths = [Path(src).resolve() for src in sources]
+        if recursive:
+            await self._upload_recursively(src_paths, basedir, dst_dir, chunk_size, address_map)
+            return
+        await self._upload_files(src_paths, basedir, dst_dir, chunk_size, address_map)
```

Comment:
I think the following pattern is more readable, as this method does not have any return value.
```suggestion
        else:
            await self._upload_files(src_paths, basedir, dst_dir, chunk_size, address_map)
```